File: webapp/render.py
"""Общие функции отображения результата расчёта — используются на
странице диалогового расчёта и на дашборде, чтобы детализация ИУ
выглядела одинаково везде."""

from __future__ import annotations

import altair as alt
import pandas as pd
import streamlit as st

from vulnerability_index.models import IndexResult


def render_summary(result: IndexResult, methodology) -> None:
    col1, col2, col3 = st.columns(3)
    col1.metric(
        "Индекс уязвимости (ИУ)", f"{result.value:.3f}",
        help=f"Диапазон {methodology.value_range[0]:g}–{methodology.value_range[1]:g}",
    )
    col2.metric("Категория уязвимости", result.category or "не определена")
    col3.metric("Статус верификации", result.verification_status.value)
    if result.has_assumptions:
        st.info(
            "⚠️ В расчёте использованы допущенные (не первичные) значения по "
            "отдельным показателям — см. детализацию ниже.",
            icon="⚠️",
        )
    if any(i.is_missing for c in result.components for i in c.indicators):
        st.error(
            "⚠️ По части показателей нет данных — их вес перенормирован среди "
            "остальных показателей компонента. Рекомендуется дособрать данные.",
            icon="⚠️",
        )
    if not methodology.categories_approved():
        st.caption(
            "Пороговые значения категорий уязвимости пока не утверждены "
            "уполномоченным органом — показанная категория рассчитана по "
            "временным значениям (см. страницу «Методика»)."
        )


def render_component_chart(result: IndexResult, value_range: tuple[float, float] = (0.0, 3.0)) -> None:
    order = [c.key for c in result.components]
    comp_df = pd.DataFrame(
        [
            {
                "Компонент": c.key,
                "Название": c.title,
                "Значение": c.value if c.value is not None else 0.0,
                "Есть данные": c.value is not None,
            }
            for c in result.components
        ]
    )
    chart = (
        alt.Chart(comp_df)
        .mark_bar()
        .encode(
            x=alt.X("Компонент:N", sort=order, title=None),
            y=alt.Y("Значение:Q", scale=alt.Scale(domain=list(value_range)), title="Балл компонента"),
            color=alt.condition(
                alt.datum["Есть данные"], alt.value("#2b6cb0"), alt.value("#a0aec0")
            ),
            tooltip=[
                alt.Tooltip("Название:N", title="Компонент"),
                alt.Tooltip("Значение:Q", title="Значение", format=".3f"),
            ],
        )
        .properties(height=280)
    )
    st.altair_chart(chart, width="stretch")


def render_breakdown(result: IndexResult) -> None:
    for component in result.components:
        value_str = f"{component.value:.3f}" if component.value is not None else "нет данных"
        with st.expander(f"{component.key} — {component.title} (вес {component.weight:g}): {value_str}"):
            if not component.indicators:
                st.caption("Нет собранных показателей.")
                continue
            rows = [
                {
                    "Показатель": indicator.title,
                    "Значение": indicator.data.raw_value,
                    "Балл": indicator.score,
                    "Источник": indicator.data.source_name,
                    "Особенность": (
                        "нет данных" if indicator.is_missing
                        else ("допущение" if indicator.data.is_assumption else "")
                    ),
                    "Комментарий": indicator.data.comment or "",
                }
                for indicator in component.indicators
            ]
            st.dataframe(pd.DataFrame(rows), width="stretch", hide_index=True)


def result_to_export_rows(result: IndexResult) -> list[dict]:
    rows = []
    for component in result.components:
        for indicator in component.indicators:
            rows.append(
                {
                    "Населённый пункт": result.settlement.name,
                    "Код": result.settlement.code,
                    "Компонент": component.key,
                    "Показатель": indicator.title,
                    "Значение": indicator.data.raw_value,
                    "Балл": indicator.score,
                    "Источник": indicator.data.source_name,
                    "Допущение": indicator.data.is_assumption,
                    "Дата получения": indicator.data.obtained_at.isoformat(),
                    "Значение компонента": component.value,
                    "ИУ": result.value,
                    "Категория": result.category,
                    "Версия методики": result.methodology_version,
                    "Дата расчёта": result.calculated_at.isoformat(),
                }
            )
    return rows

File: netlify-site/js/render.js
/**
 * Общие функции отображения результата расчёта — аналог `webapp/render.py`
 * из полной (Streamlit) версии, но на клиентском JS + Chart.js.
 */

function esc(value) {
  if (value === null || value === undefined) return '';
  return String(value).replace(/[&<>"']/g, (c) => ({
    '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;', "'": '&#39;',
  }[c]));
}

function fmtScore(v) {
  return v === null || v === undefined ? '—' : String(v);
}

function renderSummary(container, result, methodology) {
  const parts = [];
  parts.push('<div class="metrics-row">');
  parts.push(metricCard('Индекс уязвимости (ИУ)', result.value.toFixed(3),
    `Диапазон ${methodology.valueRange[0]}–${methodology.valueRange[1]}`));
  parts.push(metricCard('Категория уязвимости', result.category || 'не определена'));
  parts.push(metricCard('Статус верификации', result.verificationStatus || 'не верифицировано'));
  parts.push('</div>');

  if (result.hasAssumptions) {
    parts.push('<div class="alert alert-warning">⚠️ В расчёте использованы допущенные (не первичные) значения по отдельным показателям — см. детализацию ниже.</div>');
  }
  if (result.hasMissing) {
    parts.push('<div class="alert alert-error">⚠️ По части показателей нет данных — их вес перенормирован среди остальных показателей компонента. Рекомендуется дособрать данные.</div>');
  }
  if (!methodology.categoriesApproved()) {
    parts.push('<p class="caption">Пороговые значения категорий уязвимости пока не утверждены уполномоченным органом — показанная категория рассчитана по временным значениям (см. раздел «Методика»).</p>');
  }
  container.innerHTML = parts.join('\n');
}

function metricCard(label, value, help) {
  return `<div class="metric">
    <div class="metric-label">${esc(label)}</div>
    <div class="metric-value">${esc(value)}</div>
    ${help ? `<div class="metric-help">${esc(help)}</div>` : ''}
  </div>`;
}

function renderComponentChart(canvas, result, valueRange) {
  const labels = result.components.map((c) => c.key);
  const values = result.components.map((c) => (c.value !== null ? Number(c.value.toFixed(3)) : 0));
  const colors = result.components.map((c) => (c.value !== null ? '#2b6cb0' : '#a0aec0'));
  const titles = result.components.map((c) => c.title);

  if (canvas._chartInstance) canvas._chartInstance.destroy();
  canvas._chartInstance = new Chart(canvas, {
    type: 'bar',
    data: { labels, datasets: [{ label: 'Балл компонента', data: values, backgroundColor: colors }] },
    options: {
      scales: { y: { min: valueRange[0], max: valueRange[1], title: { display: true, text: 'Балл компонента' } } },
      plugins: {
        legend: { display: false },
        tooltip: { callbacks: { title: (items) => titles[items[0].dataIndex] } },
      },
    },
  });
}

function renderBreakdown(container, result) {
  container.innerHTML = result.components.map((c) => {
    const valueStr = c.value !== null ? c.value.toFixed(3) : 'нет данных';
    const rows = c.indicators.length
      ? c.indicators.map((i) => `
        <tr>
          <td>${esc(i.title)}</td>
          <td>${esc(i.rawValue)}</td>
          <td>${fmtScore(i.score)}</td>
          <td>${esc(i.sourceName)}</td>
          <td>${i.isMissing ? 'нет данных' : (i.isAssumption ? 'допущение' : '')}</td>
          <td>${esc(i.comment)}</td>
        </tr>`).join('')
      : '<tr><td colspan="6">Нет собранных показателей.</td></tr>';
    return `<details class="component-details">
      <summary>${esc(c.key)} — ${esc(c.title)} (вес ${c.weight}): ${valueStr}</summary>
      <table class="data-table">
        <thead><tr><th>Показатель</th><th>Значение</th><th>Балл</th><th>Источник</th><th>Особенность</th><th>Комментарий</th></tr></thead>
        <tbody>${rows}</tbody>
      </table>
    </details>`;
  }).join('\n');
}

function downloadCSV(filename, rows) {
  if (!rows.length) return;
  const headers = Object.keys(rows[0]);
  const escCsv = (v) => {
    const s = v === null || v === undefined ? '' : String(v);
    return /[",\n;]/.test(s) ? `"${s.replace(/"/g, '""')}"` : s;
  };
  const csv = [headers.join(';'), ...rows.map((r) => headers.map((h) => escCsv(r[h])).join(';'))].join('\n');
  const blob = new Blob(['\uFEFF' + csv], { type: 'text/csv;charset=utf-8;' });
  const url = URL.createObjectURL(blob);
  const a = document.createElement('a');
  a.href = url;
  a.download = filename;
  a.click();
  URL.revokeObjectURL(url);
}

window.RenderUtils = { esc, fmtScore, renderSummary, renderComponentChart, renderBreakdown, downloadCSV, metricCard };
